### DISCUSSION TITLE
Automatic thumbnail generation

### DISCUSSION THREAD

**Discussion by unleashit** [directus/directus#16448] [Archive]
Hi, I'm new to Directus and loving it so far. I haven't spent much time with GUIs in a while and it's amazing to see how far things have come since having to hack around pages and posts, etc.

Anyway I've read the docs and searched the discussions but I can't seem to figure out how (or if) it might be possible to define the default thumbnail sizes and types for a given image field. And/or when uploading directly to the file manager. I notice that when I upload through the file manager, I get two additional files for each file uploaded (one cropped and one compressed).

I understand the the philosophy here is more Cloudendary style... i.e. set the properties from HTML and have the server generate/cache which is pretty cool. But how can this work in the context of external storage like S3 and a CDN where the public URL isn't the same as Directus? It doesn't seem like it can, so it seems like either I have to manually generate and upload all thumbs in advance (and still get 3 additional files each which can be costly), or persuade Directus to generate them the right way on upload. Is there a way I've missed? I don't mind creating a hook if someone can give me some pointers for how/where to intercept.

Thanks in advance.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for the kind words <3

> Anyway I've read the docs and searched the discussions but I can't seem to figure out how (or if) it might be possible to define the default thumbnail sizes and types for a given image field

The way it currently works is that you can define thumbnail presets from the global project settings, and then request the thumbnail in one of those presets with the `key` query parameter: `/assets/<uuid>?key=homepage-hero`. You could add a dropdown field to your collection to choose the preferred preset in your own usage, but that's not a system feature.

> And/or when uploading directly to the file manager

Not entirely sure what you mean by this, but thumbnails are generated when they're first requested, and then served from the pre-generated size on subsequent requests.

> I notice that when I upload through the file manager, I get two additional files for each file uploaded (one cropped and one compressed).

The additionally generated files you're seeing are thumbnails requested by the Directus app. These aren't generated on upload, but on first request of that thumbnail size. The app will effectively immediately request that, as the newly uploaded file has a thumbnail in the file library.

> I understand the the philosophy here is more Cloudendary style... i.e. set the properties from HTML and have the server generate/cache which is pretty cool. But how can this work in the context of external storage like S3 and a CDN where the public URL isn't the same as Directus? It doesn't seem like it can, so it seems like either I have to manually generate and upload all thumbs in advance (and still get 3 additional files each which can be costly), or persuade Directus to generate them the right way on upload. Is there a way I've missed? I don't mind creating a hook if someone can give me some pointers for how/where to intercept.

The assets endpoint will stream existing files directly from S3, so the idea here is that you'd point your CDN to the Directus assets endpoint rather than relying on S3 to publicly host the files. If you were to host all files publicly through the storage provider directly, you'd also lose all access control that Directus adds to files.

**Comment by unleashit** [comment]
No doubt it could win on capabilities too, but it should definitely be nominated for he sexiest headless CMS award :p

Thanks, I really appreciate the clarifications. Since this is just a personal project, my plan was to keep the front end static and only expose Directus for admin and at build time when generating the static files. I even have Directus on a Sqlite database since that would be plenty for the use case.... which is just a CMS and doesn't require private files or anything more than read access on the front end.

Since the site is pretty simple I think I can manage with separate fields for thumbs and pre-generating them. But it would be nice to be able to have control over generation at upload time for other use cases where you might not want Directus responsible for asset serving. Maybe for a site with a lot of frequent user uploads, having a dedicated asset service would be easier/less costly then scaling Directus. Or make it easier to work with Serverless/no local file system situations. I haven't set this up personally, but I think ACL issues could be solved by appending to the file name (not sure if there's a universal convention or it would need to be couple with an edge worker)?

Not a major gripe though and in most use cases you're going to be hitting the API constantly, so why not also hit the asset API.

>     And/or when uploading directly to the file manager
> 
> Not entirely sure what you mean by this, but thumbnails are generated when they're first requested, and then served from the pre-generated size on subsequent requests.

Just meant when manually uploading images in bulk to the file manager in the admin (with the plan to later assign them to content). Ok, that makes sense. That's fine for my case actually, as the total number of images are pretty low. They're do look great in the UI :)

**Reply by rijkvanzanten** [reply]
> but I think ACL issues could be solved by appending to the file name (not sure if there's a universal convention or it would need to be couple with an edge worker)?

This is really where the discussion starts 😄 By relying on a third party (static) host, Directus isn't able to execute any logic, and therefore can't check for permissions. In that case we either have to rely on security by obfuscation with some file naming structure that's hard to guess, or rely on whatever authorization capabilities the third party might have (for example S3 has the concept of "signed URLs" for temporary file access), to that's hard to implement across multiple providers, as they all work differently and might not even support this semi-private access 🤔

**Comment by unleashit** [comment]
Yeah, I noticed this issue is discussed elsewhere so I don't mind leaving it over there. My personal opinion is that Directus shouldn't have to be responsible for file security if the developer chooses to take it into their own hands. No doubt you guys have plenty to do. I'm bigger on front end then back end, so I'm probably not aware of all the implications of some of these things... but it "seems" like if Directus could add in some form the required role to the file name, the developer should then be able to create a service that authorizes and sends a rewritten url (signed url in s3 for example) to the client. My understanding of this is that the small risk (user publishes or sends the URL somewhere else before the expiration) is currently considered more or less acceptable. But your mileage on that opinion may vary because again, my backend experience is more on the traditional side.
